demos/monge_ampere_helmholtz.py

```diff
@@ -0,0 +1,208 @@
+# this numerical error. A good indicator for where resolution is required
+# is to look at the curvature of the solution which can be expressed
+# in terms of the norm of the Hessian. A monitor function
+# that targets high resolution in places of high curvature then looks like
+#
+# .. math::
+#
+#    m = 1 + \alpha \frac{H(u_h):H(u_h)}{\max_{{\bf x}\in\Omega} H(u):H(u)}
+#
+# where :math:`:` indicates the inner product, i.e. :math:`\sqrt{H:H}` is the Frobenius norm
+# of :math:`H`. We have normalized such that the minimum of the monitor function is one (where
+# the error is zero), and its maximum is :math:`1 + \alpha` (where the curvature is maximal). This
+# means we can select the ratio between the largest area and smallest area triangle in the
+# moved mesh as :math:`1+\alpha`.
+#
+# In the following implementation we use the exact solution :math:`u_{\text{exact}}` which we
+# have as a symbolic UFL expression, and thus we can also obtain the Hessian symbolically as
+# :code:`grad(grad(u_exact))`. To compute its maximum norm however we do interpolate it
+# to a P1 function space `V` and take the maximum of the array of nodal values.
+
+alpha = Constant(5.0)
+
+
+def monitor(mesh):
+    V = FunctionSpace(mesh, "CG", 1)
+    Hnorm = Function(V, name="Hnorm")
+    H = grad(grad(u_exact(mesh)))
+    Hnorm.interpolate(sqrt(inner(H, H)))
+    Hnorm_max = Hnorm.dat.data.max()
+    m = 1 + alpha * Hnorm / Hnorm_max
+    return m
+
+
+fig, axes = plt.subplots()
+m = Function(u_h, name="monitor")
+m.interpolate(monitor(mesh))
+contours = tripcolor(m, axes=axes)
+fig.colorbar(contours)
+plt.savefig("monge_ampere_helmholtz-monitor.jpg")
+fig, axes = plt.subplots()
+
+# .. figure:: monge_ampere_helmholtz-monitor.jpg
+#    :figwidth: 60%
+#    :align: center
+#
+# As in the `previous Monge-Ampère demo <./monge_ampere1.py.html>`__, we use the
+# MongeAmpereMover to perform the mesh movement based on this monitor. We need
+# to provide the monitor as a callback function that takes the mesh as its
+# input just as we have defined it above. During the iterations of the mesh
+# movement process the monitor will then be re-evaluated in the (iteratively)
+# moved mesh nodes so that, as we improve the mesh, we can also more accurately
+# express the monitor function in the desired high-resolution areas.
+
+mover = MongeAmpereMover(mesh, monitor, method="quasi_newton")
+mover.move()
+
+# Plotting the resulting mesh
+
+from firedrake.pyplot import triplot
+
+fig, axes = plt.subplots()
+triplot(mover.mesh, axes=axes)
+axes.set_aspect(1)
+plt.savefig("monge_ampere_helmholtz-adapted_mesh.jpg")
+
+# .. figure:: monge_ampere_helmholtz-adapted_mesh.jpg
+#    :figwidth: 60%
+#    :align: center
+#
+# Now let us see whether the numerical error has actually been reduced
+# if we solve the PDE on this mesh:
+
+u_h = solve_helmholtz(mover.mesh)
+error = u_h - u_exact(mover.mesh)
+print("L2-norm error on moved mesh:", sqrt(assemble(dot(error, error) * dx)))
```

Comment:
Again, I think it'd be nice for the printed error value to actually appear in the html.